File: apps/mobile/src/modules/photo-viewer/photoReactionTally.ts
import type { PhotoReaction } from './photoReactionState'

export const REACTION_MERGE_WINDOW_MS = 800

export interface PhotoReactionTally {
  count: number
  deadline: number
  reaction: PhotoReaction
}

export interface PhotoReactionFlush {
  count: number
  reaction: PhotoReaction
}

export interface PhotoReactionTallyStep {
  flush: PhotoReactionFlush | null
  tally: PhotoReactionTally | null
}

function toFlush(tally: PhotoReactionTally | null): PhotoReactionFlush | null {
  return tally ? { count: tally.count, reaction: tally.reaction } : null
}

export function accumulateReactionTally(
  current: PhotoReactionTally | null,
  reaction: PhotoReaction,
  count: number,
  now: number,
  windowMs: number = REACTION_MERGE_WINDOW_MS,
): PhotoReactionTallyStep {
  if (count <= 0) {
    return { flush: null, tally: current }
  }

  const extendsCurrent = current !== null && current.reaction === reaction && now < current.deadline
  return {
    flush: extendsCurrent ? null : toFlush(current),
    tally: {
      count: extendsCurrent ? current.count + count : count,
      deadline: now + windowMs,
      reaction,
    },
  }
}

export function expireReactionTally(current: PhotoReactionTally | null, now: number): PhotoReactionTallyStep {
  if (current === null || now < current.deadline) {
    return { flush: null, tally: current }
  }
  return { flush: toFlush(current), tally: null }
}

export function drainReactionTally(current: PhotoReactionTally | null): PhotoReactionTallyStep {
  return { flush: toFlush(current), tally: null }
}

File: apps/mobile/modules/photo-masonry/ios/Models/PhotoReactionTally.swift
import Foundation

struct PhotoReactionTally: Equatable, Sendable {
  let count: Int
  let deadline: TimeInterval
  let reaction: String
}

struct PhotoReactionFlush: Equatable, Sendable {
  let count: Int
  let reaction: String
}

struct PhotoReactionTallyStep: Equatable, Sendable {
  let flush: PhotoReactionFlush?
  let tally: PhotoReactionTally?
}

enum PhotoReactionTallyEngine {
  static let mergeWindow: TimeInterval = 0.8

  static func accumulate(
    current: PhotoReactionTally?,
    reaction: String,
    count: Int,
    now: TimeInterval,
    window: TimeInterval = mergeWindow
  ) -> PhotoReactionTallyStep {
    guard count > 0 else { return PhotoReactionTallyStep(flush: nil, tally: current) }
    let extendsCurrent = current?.reaction == reaction && now < (current?.deadline ?? 0)
    return PhotoReactionTallyStep(
      flush: extendsCurrent ? nil : current.map { PhotoReactionFlush(count: $0.count, reaction: $0.reaction) },
      tally: PhotoReactionTally(
        count: extendsCurrent ? (current?.count ?? 0) + count : count,
        deadline: now + window,
        reaction: reaction
      )
    )
  }

  static func expire(current: PhotoReactionTally?, now: TimeInterval) -> PhotoReactionTallyStep {
    guard let current, now >= current.deadline else {
      return PhotoReactionTallyStep(flush: nil, tally: current)
    }
    return PhotoReactionTallyStep(
      flush: PhotoReactionFlush(count: current.count, reaction: current.reaction),
      tally: nil
    )
  }

  static func drain(current: PhotoReactionTally?) -> PhotoReactionTallyStep {
    PhotoReactionTallyStep(
      flush: current.map { PhotoReactionFlush(count: $0.count, reaction: $0.reaction) },
      tally: nil
    )
  }
}
